### ISSUE TITLE
Batch Normalization for 3D data

### ISSUE THREAD

**Issue by SSSuperMac** [Lasagne/Lasagne#822]
I used the `lasagne.layers.BatchNormLayer` to deal with 3D data (3D Volume data), but  I encountered some problems when the theano compiled the `update` function. some errors are like below:
```
   updates = lasagne.updates.momentum(reg_loss, params, learning_rate, cfg['momentum'])
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/lasagne/updates.py", line 259, in momentum
    updates = sgd(loss_or_grads, params, learning_rate)
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/lasagne/updates.py", line 167, in sgd
    grads = get_or_compute_grads(loss_or_grads, params)
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/lasagne/updates.py", line 143, in get_or_compute_grads
    return theano.grad(loss_or_grads, params)
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/theano/gradient.py", line 539, in grad
    handle_disconnected(elem)
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/theano/gradient.py", line 526, in handle_disconnected
    raise DisconnectedInputError(message)
theano.gradient.DisconnectedInputError:
Backtrace when that variable is created:

  File "myTrain.py", line 183, in <module>
    main(args)
  File "myTrain.py", line 124, in main
    model = config_module.get_model()
  File "testNet.py", line 51, in get_model
    name = 'bn1')
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/lasagne/layers/normalization.py", line 262, in __init__
    trainable=False, regularizable=False)
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/lasagne/layers/base.py", line 234, in add_param
    param = utils.create_param(spec, shape, name)
  File "/home/machao/anaconda3/envs/python27/lib/python2.7/site-packages/lasagne/utils.py", line 381, in create_param
    spec = theano.shared(spec, broadcastable=bcast)
```
I had tested this layer in the 2D images for object recognition, and it acted normally, so I am wondering if I have missed something important about this function, or the `BatchNormLayer` doesn't supported the 3D data input at present?

**Comment by SSSuperMac** [comment]
I have worked it out!

**Comment by f0k** [comment]
> I have worked it out!

Good! What was the problem, in case anybody else googles for this error? I assume you forgot the `trainable=True` flag in your `params = lasagne.layers.get_all_params(network, trainable=True)` call, when collecting the parameters for the update dictionary. This is needed so `get_all_params` only returns the parameters that participate in the forward pass. Without filtering for the `trainable` tag, it will also return parameters that are only important at prediction time (in this case, the exponential running averages over means and variances). This is independent of the dimensionality of the data you batch-normalize.

**Comment by SSSuperMac** [comment]
@f0k 
yeah, that's it! the `trainable =True` was missed, then theano tried to get the gradients of `mean` and `inv_std` in the BatchNormLayer, but both were not in the grad_dict, so the error occurred!
